mkosi/distributions/__init__.py

```diff
@@ -0,0 +1,95 @@
+# SPDX-License-Identifier: LGPL-2.1+
+
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Set
+
+from ..backend import Distribution, MkosiArgs, PartitionIdentifier
+
+DRACUT_SYSTEMD_EXTRAS = [
+    "/usr/bin/systemd-repart",
+    "/usr/lib/systemd/system-generators/systemd-veritysetup-generator",
+    "/usr/lib/systemd/system/initrd-root-fs.target.wants/systemd-repart.service",
+    "/usr/lib/systemd/system/initrd-usr-fs.target",
+    "/usr/lib/systemd/system/sysinit.target.wants/veritysetup.target",
+    "/usr/lib/systemd/system/systemd-repart.service",
+    "/usr/lib/systemd/system/systemd-volatile-root.service",
+    "/usr/lib/systemd/system/veritysetup.target",
+    "/usr/lib/systemd/systemd-veritysetup",
+    "/usr/lib/systemd/systemd-volatile-root",
+    "/usr/bin/systemd-ask-password",
+    "/usr/bin/systemd-tty-ask-password-agent"
+]
+
+
+def configure_dracut(args: MkosiArgs, packages: Set[str], root: Path) -> None:
+    if "dracut" not in packages:
+        return
+
+    dracut_dir = root / "etc/dracut.conf.d"
+    dracut_dir.mkdir(mode=0o755)
+
+    dracut_dir.joinpath('30-mkosi-hostonly.conf').write_text(
+        f'hostonly={"yes" if args.hostonly_initrd else "no"}\n'
+        'hostonly_default_device=no\n'
+    )
+
+    dracut_dir.joinpath("30-mkosi-qemu.conf").write_text('add_dracutmodules+=" qemu "\n')
+
+    with dracut_dir.joinpath("30-mkosi-systemd-extras.conf").open("w") as f:
+        for extra in DRACUT_SYSTEMD_EXTRAS:
+            f.write(f'install_optional_items+=" {extra} "\n')
+
+    if args.hostonly_initrd:
+        dracut_dir.joinpath("30-mkosi-filesystem.conf").write_text(
+            f'filesystems+=" {(args.output_format.needed_kernel_module())} "\n'
+        )
+
+    if args.get_partition(PartitionIdentifier.esp):
+        # These distros need uefi_stub configured explicitly for dracut to find the systemd-boot uefi stub.
+        if args.distribution in (Distribution.ubuntu,
+                                 Distribution.debian,
+                                 Distribution.mageia,
+                                 Distribution.openmandriva,
+                                 Distribution.gentoo):
+            dracut_dir.joinpath("30-mkosi-uefi-stub.conf").write_text(
+                "uefi_stub=/usr/lib/systemd/boot/efi/linuxx64.efi.stub\n"
+            )
+
+        # efivarfs must be present in order to GPT root discovery work
+        dracut_dir.joinpath("30-mkosi-efivarfs.conf").write_text(
+            '[[ $(modinfo -k "$kernel" -F filename efivarfs 2>/dev/null) == /* ]] && add_drivers+=" efivarfs "\n'
+        )
+
+
+class DistributionInstaller(MkosiArgs):
+    def hook_install_etc_locale(self, root: Path, cached: bool) -> None:
+        pass
```

Comment:
That's definitely a possibility, too, but I'm not sure whether cramming more things into the usually already long `install_foo` functions is the best way, since where we need to look where this might be touched will be less strictly localised. 

I prefer more hooks (preferably at the end) of functions like `install_etc_locale` that are then noops for the majority than moving things into the install functions.

Reply:
I definitely think having fewer hooks trumps having more hooks, especially if the hook is only required because 1/2 distros need to do something weird due.

I think our goal should be to get rid of all hooks except install and the package manager specific stuff. 

If we ever want to expose a public API to implement out of tree distro support via this, we want to keep it as minimal as possible, since what I've seen over the years maintaining mkosi is that things can change very frequently, and any hooks we have now might very well not be needed anymore in the near future, so having one generic hook where we do as much as possible seems a lot better than many specific hooks that might become irrelevant in newer versions.

Reply:
It's funny how two people will arrive at two completely different results from the same starting position. :)

When I'm saying that more hooks are better than fewer hooks, then I mean: within reason. I would define that as leaving it on the table of hooking the steps like `install_etc_locale` that we already have (or add in the future). I think this is preferable for one big reason: The logic stays localised and it's easier to reason about what state the installation will be at each step, so ordering problems become less likely. 

I'm also not advocating for committing to some public API today. If we drop certain things or hooks become unnecessary, that's fine. Just for the problem at hand (just a split) I think it's easier to port the existing logic as faithfully as possible.

Reply:
I still thinks this sets the wrong precedent. Having few hooks and fewer distro specific hooks will establish clearly that we want to minimize distro specific behavior. Codifying workarounds in a custom hook function is almost like saying we accept the workaround, whereas we want for every non-standard addition to stick out like a sore thumb instead. We'll achieve that much better by only supporting as few hooks as we can. Yes, this does make some stuff less localized, but I think overall this is still a good trade off.